feat(dot/parachain): create worker pool for PVF host

edwardmack opened the pull request:
## Changes
This creates a PVF host which will be called for candidate validation.  It enables the use of a worker pool that will contain workers that will preform the candidate validation task.  

- The candidate validation subsystem creates and starts an instance of the PVF Host.  
- The PVF host contains a pool of candidate validation workers (one for each of the validation code instances)
- The candidate validation subsystem calls the Validate method of the PVF host to validate candidates as needed.
- The PVF host initializes validation workers when needed


<!-- Brief list of functional changes -->

## Tests

<!-- Detail how to run relevant tests to the changes -->

```sh
go test github.com/ChainSafe/gossamer/dot/parachain/candidate-validation
```

## Issues
closes #3934 

<!-- Write the issue number(s), for example: #123 -->

edwardmack committed:
add pvf worker pool skeleton

edwardmack committed:
add test to validation host

edwardmack committed:
create worker pool skeleton for PVF host

edwardmack committed:
add validation logic to workers

edwardmack committed:
address lint issues.

edwardmack committed:
add functionality for validate from chainstate to pvf host and tests

edwardmack committed:
add updated mock files

edwardmack committed:
clean-up unused comments

EclesioMeloJunior commented:
Hey @edwardmack could you explain how the `candidate validation workers` are instantiated? what is the logic to start a new worker?

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_pvf_host_base

edwardmack commented:
> Hey @edwardmack could you explain how the `candidate validation workers` are instantiated? what is the logic to start a new worker?

Hi @EclesioMeloJunior when a call to host.Validate is made the validate function calls `poolContainsWorker` to determine if there is a worker with the given code hash, if not, then one is created by calling `workerPool.newValidationWorker`.

edwardmack committed:
address merge conflicts

kishansagathiya commented on dot/parachain/candidate-validation/mocks_generate_test.go:
```diff
@@ -4,4 +4,5 @@
 //go:generate mockgen -destination=mocks_test.go -package=$GOPACKAGE . PoVRequestor
-//go:generate mockgen -destination=mocks_runtime_test.go -package $GOPACKAGE github.com/ChainSafe/gossamer/dot/parachain/runtime RuntimeInstance,ValidationHost
+//go:generate mockgen -destination=mocks_blockstate_test.go -package=$GOPACKAGE . BlockState
+//go:generate mockgen -destination=mocks_instance_test.go -package $GOPACKAGE github.com/ChainSafe/gossamer/lib/runtime Instance
```
I had used a subset of methods from the original `Runtime` instance in `RuntimeInstance`.

Do we need all the methods?

edwardmack replied:
In the process from chainstate test I'm mocking block state to make a mock call to GetRuntime, and mock blockState expects the return to implement the full runtime interface.  When I try without I get this error 
```
wrong type of argument 0 to Return for *candidatevalidation.MockBlockState.GetRuntime: candidatevalidation.MockRuntimeInstance is not assignable to runtime.Instance
```

kishansagathiya commented on dot/parachain/pvf/host.go:
```diff
@@ -0,0 +1,148 @@
+
+type ValidationHost struct {
+	wg     sync.WaitGroup
+	stopCh chan struct{}
```
this doesn't seem to be doing anything!

edwardmack replied:
I've removed the concurrency code that wasn't doing anything.

kishansagathiya commented on dot/parachain/pvf/host.go:
```diff
@@ -0,0 +1,148 @@
+var logger = log.NewFromGlobal(log.AddContext("pkg", "pvf"), log.SetLevel(log.Debug))
+
+type ValidationHost struct {
+	wg     sync.WaitGroup
```
No need for waitgroup here. Instead pass the waitgroup from candidate validation to the `Start` method

kishansagathiya commented on dot/parachain/pvf/host.go:
```diff
@@ -0,0 +1,148 @@
+func (v *ValidationHost) Start() {
+	v.wg.Add(1)
+	logger.Debug("Starting validation host")
+	go func() {
+		defer v.wg.Done()
+	}()
```
This defer will get executed immediately. remove it.

kishansagathiya replied:
I don't get it. `Start` doesn't do anything other than a putting a log.

kishansagathiya commented on dot/parachain/pvf/host.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+func (v *ValidationHost) Start() {
+	v.wg.Add(1)
```
We must add `wg.Add()`, before calling the goroutine we want to wait on.
Then we pass `wg` inside that goroutine. 
Then we call `wg.Done` at the end of our goroutine.

There might be some good blogs on the topic, but posting one from my search results here https://blog.logrocket.com/concurrency-patterns-golang-waitgroups-goroutines/
https://www.reddit.com/r/golang/comments/1buj596/til_dont_add_waitgroup_inside_go_routine/

EclesioMeloJunior commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -86,35 +90,30 @@ func (cv *CandidateValidation) processMessages(wg *sync.WaitGroup) {
+					PvfExecTimeoutKind:      msg.PvfExecTimeoutKind,
+					ResultCh:                taskResult,
+				}
+				go cv.pvfHost.Validate(validationTask)
+
+				result := <-taskResult
```
I didn't understand this part, is this meant to be concurrent? From my understanding this blocks the process until it receives the result, so it becomes a sync task and can be done w/o dispatching a goroutine.

```go
result := cv.pvfHost.Validate(validationTask)
```

kishansagathiya replied:
That's right. @edwardmack You should create a queue of validation tasks inside `Host`. You should just send the task there and not wait for its result.
```
				if result.InternalError != nil {
					logger.Errorf("failed to validate from exhaustive: %w", result.InternalError)
					msg.Ch <- parachaintypes.OverseerFuncRes[pvf.ValidationResult]{
						Err: result.InternalError,
					}
				} else {
					msg.Ch <- parachaintypes.OverseerFuncRes[pvf.ValidationResult]{
						Data: *result.Result,
					}
                }
```
The code above can be processed where you handle the validate task.

kishansagathiya commented on dot/parachain/pvf/host.go:
```diff
@@ -0,0 +1,148 @@
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "pvf"), log.SetLevel(log.Debug))
+
+type ValidationHost struct {
```
We could just call it `Host`

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/pvf/worker_pool.go:
```diff
@@ -0,0 +1,212 @@
+	mtx sync.RWMutex
+	wg  sync.WaitGroup
+
+	workers map[parachaintypes.ValidationCodeHash]*worker
```
A validation code can be mapped to a parachain, rather than a candidate. Multiple candidates can have the same validation code.
This would mean that we will have only one worker per parachain. Is there any benefit of dividing workers based on the parachains? Keep in mind that we should be able to validate multiple candidate from the same parachain without waiting on each other.

https://paritytech.github.io/polkadot-sdk/book/runtime-api/validation-code.html

edwardmack replied:
I'm using ValidationCodeHash as key for these workers since the worker is storing a runtime instance based on the validation code, we may update this key in the future for handling multiple instances of candidates.

kishansagathiya replied:
I am starting to wonder if two parachains can have the same validation code. I have a feeling that it is possible. In which case we will have to change this a bit. but let's evaluate and tackle that as a separate piece of work.

kishansagathiya reviewed (commented):
You can put more comments explaining code, that would help us understand the code and design better while reviewing.

kishansagathiya commented on dot/parachain/pvf/worker_pool.go:
```diff
@@ -0,0 +1,212 @@
+	allWorkersDoneCh := make(chan struct{})
+	go func() {
+		defer close(allWorkersDoneCh)
+		v.wg.Wait()
```
We use `wg.Wait` to wait on all the goroutine that we have added in the waitgroup. This means that execution will wait until anything all the goroutines are done.

But when you put `wg.Wait` inside a goroutine, you are not waiting at all. The execution will continue. wg.Wait will have not effects.

kishansagathiya commented on dot/parachain/pvf/worker_pool.go:
```diff
@@ -0,0 +1,212 @@
+	maxRequestsAllowed uint = 60
+)
+
+type validationWorkerPool struct {
```
just workerPool should be good.

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/pvf/worker_pool.go:
```diff
@@ -0,0 +1,212 @@
+	}
+}
+
+//type validationWorker struct {
+//	worker *worker
+//	queue  chan *workerTask
+//}
```
remove?

edwardmack replied:
removed

kishansagathiya commented on dot/parachain/pvf/worker.go:
```diff
@@ -0,0 +1,130 @@
+	}, nil
+}
+
+func (w *worker) run(queue chan *workerTask, wg *sync.WaitGroup) {
```
worker already has queue. We don't need to pass it.

edwardmack replied:
removed queue from run parameters

kishansagathiya commented on dot/parachain/pvf/host.go:
```diff
@@ -0,0 +1,148 @@
+	}
+	validationCodeHash := msg.ValidationCode.Hash()
+	if v.workerPool.containsWorker(validationCodeHash) {
+
```
remove the space

edwardmack replied:
removed.

kishansagathiya commented on dot/parachain/pvf/worker.go:
```diff
@@ -0,0 +1,130 @@
+		wg.Done()
+	}()
+
+	for task := range queue {
+		w.executeRequest(task)
+	}
```
You can run all executeRequest concurrently as a goroutine. Right now they are running one by one.

kishansagathiya commented on dot/parachain/pvf/worker.go:
```diff
@@ -0,0 +1,130 @@
+type worker struct {
+	workerID parachaintypes.ValidationCodeHash
+	instance *parachainruntime.Instance
+	queue    chan *workerTask
```
This queue channel has capacity of only one. A new workerTask cannot be added until the existing one finishes. Make this a buffered channel.

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_pvf_host_base

edwardmack committed:
resolve merge conflicts

edwardmack committed:
refactor ValidationHost to Host, and validationWorkerPool to workerPool

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_pvf_host_base

edwardmack committed:
refactor go concurrancy

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -38,25 +43,16 @@ type BlockState interface {
-			}
-			return
-		}
-	}
+func (cv *CandidateValidation) Run(context.Context, <-chan any) {
+	cv.wg.Add(1)
+	go cv.processMessages(&cv.wg)
 }
```
Why?

edwardmack replied:
I thought this looked cleaner, we discussed this so I've reverted it.

axaysagathiya reviewed (changes requested):
I see many unnecessary diffs. It might be you haven't resolved conflicts properly.
I highly encourage you to self-review the PR once.

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_pvf_host_base

edwardmack committed:
revert run to match interface

edwardmack committed:
remove redundant structs for simplicty, removed go routines to simplify code

edwardmack committed:
removed pvf package, simplified code removed concurrency

edwardmack committed:
added missing mock file

edwardmack commented:
I've removed the `pvf` package and moved code into `candidate-validation` package since I think that makes more sense.  Cleaned up incorrect concurrency code.  This is ready for a re-review.

edwardmack committed:
address deep source comments

EclesioMeloJunior commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,149 @@
+	return vr.ValidResult != nil
+}
+
+type ValidValidationResult struct {
```
```suggestion
type Valid struct {
```

edwardmack replied:
Good idea, updated.

EclesioMeloJunior commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,93 @@
+
+	// do isProcessed check here
+	if _, ok := w.isProcessed[candidateHash]; ok {
+		// TODO: determine what the isPreccessed check should return, and if re-trying is allowed
+		//  get a better understanding of what the isProcessed check should be checking for
+		logger.Debugf("candidate %x already processed", candidateHash)
```
should we return here?

edwardmack replied:
Yes, I've updated this to return the cached value.

EclesioMeloJunior commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,103 @@
+	encodedPoVSize := uint32(len(encodedPoV))
+
+	if encodedPoVSize > maxPoVSize {
+		ci := ParamsTooLarge
+		return &ci, nil
```
I see that the return should be a pointer to `ReasonForInvalidity`, I would like to ask if there is any reason for that? I would suggest to return the raw value itself, and then the caller uses the reference to it

edwardmack replied:
I have this returning a pointer so that the caller can check if the return value is nil, then there was no error, the basic check passes.  However it does feel clumsy to treat this as a pointer and having to de-reference this.  So, it's currently if ReasonForInvalidity is nil, then it is valid.

EclesioMeloJunior commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,103 @@
+		return &ValidationResult{InvalidResult: validationErr}, nil //nolint
+	}
+	// create worker if not in pool
+	if !v.workerPool.containsWorker(validationCodeHash) {
+		worker, err := v.workerPool.newValidationWorker(*msg.ValidationCode)
+		if err != nil {
+			return nil, err
+		}
+
+		// sanity check
+		if worker.workerID != validationCodeHash {
+			return nil, fmt.Errorf("workerID does not match validationCodeHash")
+		}
+	}
```
I think this is worker's pool responsibility (check if the worker exists, if not then create a new one as well as the sanity check)

edwardmack replied:
I've moved is logic to the worker pool.

EclesioMeloJunior commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,103 @@
+		}
+
+		// sanity check
+		if worker.workerID != validationCodeHash {
+			return nil, fmt.Errorf("workerID does not match validationCodeHash")
```
Is this sanity check really needed? what is the probability of the function `newValidationWorker` to create a worker with a worker ID different of the one passed as argument?

edwardmack replied:
Good point, I've removed this check.

edwardmack committed:
address PR comments

timwu20 commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,79 @@
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "pvf"), log.SetLevel(log.Debug))
+
+type Host struct {
```
any specific reason why this is public? Seems to be only referenced in the package.

edwardmack replied:
no reason, I've updated this to be private.

timwu20 commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,93 @@
+
+	// do isProcessed check here
+	if processed, ok := w.isProcessed[candidateHash]; ok {
+		logger.Debugf("candidate %x already processed", candidateHash)
```
should this ever happen?

edwardmack replied:
It seems that this may happen, this was done to handle the requirement from host implementer's guide "Multiple requests for preparing the same artifact are coalesced, so that the work is only done once."

timwu20 commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,93 @@
+		return processed, nil
+	}
+	validationResult, err := w.instance.ValidateBlock(task.work)
+
```
```suggestion
```

edwardmack replied:
Removed.

timwu20 commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// submitRequest given a request, the worker pool will get the worker for a given workerID
+// a channel in returned that the response will be dispatch on
```
Is this still the case? I'm not really understanding why this is called `wokerPool`.  It looks like a new worker is created for each `validationCodeHash` but it calls `syncWorker.executeRequest` at the end of this which is a synchronous function.  So I'm not sure how there would be parallel workers running.  Also I don't see any reference to any channels.

edwardmack replied:
No, I've updated the comment to better reflect what is happening.

timwu20 replied:
but the description of this function still says "a channel is returned"?

edwardmack committed:
make private functions/structs private

EclesioMeloJunior reviewed (approved):
LGTM! Great work Ed 🚀

timwu20 commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,164 @@
+		return fmt.Errorf("failed to create a new worker: %w", err)
+	}
+
+	v.workers[worker.workerID] = worker
```
shouldn't we check if the map already has this `workerID`? We could be overwriting an existing worker?

edwardmack replied:
I have a call to `containsWorker` at line 137 that checks if the worker exists.  Should I move that check into the `newValidationWorker` function?

timwu20 replied:
I think the worker retrieval/creation flow is a little odd.  Normally I would check have some `GetOrCreateWorker(parachaintypes.ValidationCode)` type method that would check first and return or create a new one.

kishansagathiya replied:
I agree with @timwu20. We should be checking here, that a worker isn't already present.

timwu20 commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,165 @@
+// submitRequest given a request, the worker pool will get the worker for a given task and submit the request
+// to the worker. The worker will execute the request and return the result. If the worker does not exist, a new worker
+// will be created and the request will be submitted to the worker.
+func (v *workerPool) submitRequest(msg *ValidationTask) (*ValidationResult, error) {
```
```suggestion
func (v *workerPool) executeRequest(msg *ValidationTask) (*ValidationResult, error) {
```
Maybe mirror the name since it's not submitting to a pool anymore, but rather will be executed synchronously.

edwardmack replied:
Good point, updated.

edwardmack committed:
refactor submitRequest to executeRequest

timwu20 commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,92 @@
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+)
+
+// TODO(ed): figure out a better name for this that describes what it does
```
Can we be sure we create some issues to track these TODOs.

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation.go:
```diff
@@ -10,24 +10,15 @@ import (
 	SubsystemToOverseer chan<- any
-	ValidationHost      parachainruntime.ValidationHost
 	BlockState          BlockState
+	pvfHost             *host
```
Add a comment to explain what is `pvfHost`

edwardmack replied:
Added comment

axaysagathiya commented on dot/parachain/candidate-validation/candidate_validation_test.go:
```diff
@@ -838,13 +458,10 @@ func TestCandidateValidation_validateFromChainState(t *testing.T) {
-			require.Equal(t, *tt.want, *got)
+
+			toSubsystem <- tt.msg
+			result := <-sender
```
You are running all the test cases parallelly, So you should not share the channels among them.
You might receive the value from another case in the `sender` channel.

edwardmack replied:
I've updated this to make a new sender channel for each test run.

axaysagathiya commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,79 @@
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "pvf"), log.SetLevel(log.Debug))
+
+type host struct {
+	workerPool *workerPool
+}
```
Can you add a comment explaining the `host` structure?

edwardmack replied:
Added

axaysagathiya commented on dot/parachain/candidate-validation/host_test.go:
```diff
@@ -0,0 +1,326 @@
+			},
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			validationError, _ := performBasicChecks(tt.args.candidate, tt.args.maxPoVSize, tt.args.pov,
```
Can we run these cases parallelly?
Also, Why do you ignore internal error?

edwardmack replied:
Added parallel to tests and added check for internal errors.

axaysagathiya reviewed (changes requested):
I have reviewed a few files and made some comments. I will look into other files later.

edwardmack committed:
address PR comments

kishansagathiya commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,80 @@
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+var logger = log.NewFromGlobal(log.AddContext("pkg", "pvf"), log.SetLevel(log.Debug))
```
You are in package `candidatevalidation`, but you are using a logger with `pkg` as `pvf`

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,80 @@
+	return v.workerPool.executeRequest(msg)
+}
+
+// performBasicChecks Does basic checks of a candidate. Provide the encoded PoV-block.
+// Returns ReasonForInvalidity and internal error if any.
```
```suggestion
// performBasicChecks does basic checks of a candidate. Provided the encoded PoV-block it returns ReasonForInvalidity and internal error if any.
```

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,80 @@
+		ci := BadSignature
+		return &ci, nil
+	}
+	return nil, nil
```
```suggestion

	return nil, nil
```

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,91 @@
+
+type worker struct {
+	workerID    parachaintypes.ValidationCodeHash
+	instance    *parachainruntime.Instance
```
```suggestion
	instance    ValidationHost
```

edwardmack replied:
Updated to newly created `ValidatorInstance` interface.

kishansagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,91 @@
+
+func newWorker(validationCode parachaintypes.ValidationCode) (*worker, error) {
+	validationRuntime, err := parachainruntime.SetupVM(validationCode)
+
+	if err != nil {
+		return nil, err
+	}
```
```suggestion
	if err != nil {
		return nil, err
	}
    
```

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,163 @@
+type ValidationResult struct {
+	ValidResult   *Valid
+	InvalidResult *ReasonForInvalidity
+}
```
let's make this a bit simpler and just name them `Valid` and `Invalid`

axaysagathiya replied:
Rename only the field name and not the `ReasonForInvalidity` type.
```
type ValidationResult struct {
	Valid   *Valid
	Invalid *ReasonForInvalidity
}
```

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// ValidationResult represents the result coming from the candidate validation subsystem.
+// Validation results can be either a Valid or InvalidValidationResult.
```
```suggestion
// Validation results can be either a valid or invalid.
```

There is no need to use exact struct names or function names in comments. In fact, we should avoid that.

edwardmack replied:
Ok, updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,163 @@
+// Validation results can be either a Valid or InvalidValidationResult.
+//
+// If the result is invalid,
+// store the reason for invalidity in the InvalidResult field of ValidationResult.
```
```suggestion
// If the result is invalid, store the reason for invalidity.
```

edwardmack replied:
updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,163 @@
+// If the result is invalid,
+// store the reason for invalidity in the InvalidResult field of ValidationResult.
+//
+// If the result is valid,
+// set the values of the ValidResult field of Valid.
```
```suggestion
// If the result is valid, store persisted validation data and candidate commitments.
```

edwardmack replied:
updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,163 @@
+	}
+}
+
+func newValidationWorkerPool() *workerPool {
```
just `newWorkerPool` should do.

edwardmack replied:
Updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker_pool.go:
```diff
@@ -0,0 +1,163 @@
+	}
+}
+
+func (v *workerPool) newValidationWorker(validationCode parachaintypes.ValidationCode) error {
```
```suggestion
func (v *workerPool) addNewWorker(validationCode parachaintypes.ValidationCode) error {
```

edwardmack replied:
updated.

kishansagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+func newWorker(validationCode parachaintypes.ValidationCode) (*worker, error) {
+	validationRuntime, err := parachainruntime.SetupVM(validationCode)
```
rename this to prachainRuntime

edwardmack replied:
updated.

axaysagathiya commented on dot/parachain/candidate-validation/host.go:
```diff
@@ -0,0 +1,80 @@
+
+func (v *host) validate(msg *ValidationTask) (*ValidationResult, error) {
+	validationCodeHash := msg.ValidationCode.Hash()
+	// performBasicChecks
```
This comment seems unnecessary as this is the function name.
We can remove this comment.

edwardmack replied:
removed.

axaysagathiya commented on dot/parachain/candidate-validation/host_test.go:
```diff
@@ -0,0 +1,330 @@
+			if tt.expectedError != nil {
+				require.EqualError(t, validationError, tt.expectedError.Error())
+			} else {
+				require.Nil(t, validationError)
```
Maybe you can use `require.NoError(t, validationError)`, which is the same as you used above for internal errors. This is just another perspective.

You can choose whether you want to update this or not.

edwardmack replied:
Good point, I've updated for consistency.

edwardmack replied:
When I updated this to roquire.NoError the test failed with the following error:
```
      	Error Trace:	/home/emack/projects/ChainSafe/gossamer/dot/parachain/candidate-validation/host_test.go:328
        	Error:      	Received unexpected error:
        	            	<nil>
        	Test:       	TestHost_performBasicChecks/happy_path
```
so I've changed it back.

axaysagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,91 @@
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+)
+
+type worker struct {
```
Can you add a comment about what a worker is?

edwardmack replied:
added.

axaysagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,91 @@
+)
+
+type worker struct {
+	workerID    parachaintypes.ValidationCodeHash
```
Isn't it that you were planning to change the type from `ValidationCodeHash` to parachain-id?

edwardmack replied:
There was a comment about changing this to parachain id, but I'd like to keep it as validation code hash since the validation is for that give instance of the code.  I'm concerned that there may be future parachain runtime updates, where parachain ID would stay the same, but the validation code may have changed.

axaysagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,91 @@
+		return nil, err
+	}
+
+	// do isProcessed check here
```
do we still need this comment?

edwardmack replied:
removed.

edwardmack committed:
address PR comments, rename variables and functions

edwardmack commented:
@kishansagathiya and @axaysagathiya, I think I've addressed all comments, so they can be resolved, and this is ready for a re-review.  If you still find changes that need to be done before approval please let me know.

edwardmack committed:
fix test

edwardmack committed:
remove printf from debugging

kishansagathiya commented on dot/parachain/runtime/instance.go:
```diff
@@ -89,8 +89,8 @@ func (in *Instance) ValidateBlock(params ValidationParameters) (
 }
 
-type ValidationHost interface {
-	// ValidateBlock validates a block by calling parachain runtime's validate_block call and returns the result.
```
no need to remove the comment

edwardmack replied:
re-added comment.

kishansagathiya commented on dot/parachain/candidate-validation/worker.go:
```diff
@@ -0,0 +1,92 @@
+
+func newWorker(validationCode parachaintypes.ValidationCode) (*worker, error) {
+	parachainRuntime, err := parachainruntime.SetupVM(validationCode)
+
```
remove this space

edwardmack replied:
removed.

kishansagathiya reviewed (approved):
🎉 
I have some small comments. Approving to avoid one more iteration. So, you can resolve those yourself and merge this PR.

kishansagathiya reviewed (commented):
All the concurrency related code has been removed from this PR. So, this is basically same as previously in terms of behaviour but the code has been moved around.

If we were to merge this we need to have a separate issue to actually put concurrency in this an make pvf host functional.

I want to get a nod from @timwu20 and @EclesioMeloJunior on whether you are okay to merge this in. I am asking because I am bit confused on this.

edwardmack committed:
address PR comments

edwardmack committed:
Merge branch 'feat/parachain' into ed/feat/candidate_validation_pvf_host_base

edwardmack commented:
> If we were to merge this we need to have a separate issue to actually put concurrency in this an make pvf host functional.

Issue #4185 has been created to track this issue.
